test/functional/p2p_peer_logging.py

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test node peer logging capability.
+
+Additionally, the output of contrib/peer-logging/peer-logging-parser.py should be verified manually.
+"""
+
+import glob
+from io import BytesIO
+import os
+
+from test_framework.mininode import P2PDataStore
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+TIME_SIZE = 8
+LENGTH_SIZE = 4
+COMMAND_SIZE = 12
+
+def mini_parser(dat_file):
+    """Parse a data file created by LogMessage.
+
+    From the data file we'll only check the structure.
+
+    We won't care about things like:
+    - Deserializing the payload of the message
+        - This is managed by the deserialize methods in test_framework.messages
+    - The order of the messages
+        - There's no reason why we can't, say, change the order of the messages in the handshake
+    - Message Type
+        - We can add new message types
+
+    We're ignoring these because they're simply too brittle to test here.
+    """
+    with open(dat_file, 'rb') as f_in:
```

Comment:
Instead of reimplementing the parser here, couldn't this just call `contrib/peer-logging/peer-logging-parser.py {dat_file}`?

Reply:
+1 on code deduplication. The included timestamps check (all timestamps are past genesis block) can be done on the result rather than directly in the parser.

Reply:
I think it's much better not to assume that the contrib directory is available. Functional tests can be run when the build directory is out of tree and the source directory is not available. It makes sense to me to have a limited parser in the test case rather than importing from somewhere else.

Reply:
I agree that the dependency on the contrib directory is a drawback, but then on the other hand we already have a test importing from contrib: `feature_loadblock.py` (uses `contrib/linearize` scripts). Is there really a need to run functional tests without the other sources available? They are all part of the same repository.

Reply:
> Is there really a need to run functional tests without the other sources available?

Yes, it's an explicit usage pattern to build out of tree and be able to run tests separately from the source directory.

Reply:
Hm, so the usage pattern in this case is "copy the `test/functional` folder somewhere" and executing from there, decoupled from the repository, should still work? Still not convinced why anyone should ever want/do that (maybe I'm missing something there).

Reply:
Search for "out of tree builds" in this repo. You'll see lots of issues/PRs for running tests out of tree.

> I agree that the dependency on the contrib directory is a drawback, but then on the other hand we already have a test importing from contrib: feature_loadblock.py (uses contrib/linearize scripts).

Yes, and I think that's a terrible pattern that we shouldn't repeat. Importing code from contrib essentially makes those files part of the test framework.

Reply:
To my understanding the primary point of out of tree builds is separation between the _build artifacts_ (i.e. everything spit out by `make`, like binaries, manpages etc.) and the source folder. I guess our disagreement stems from the fact that in your view, `contrib` is part of the source folder, why in my view it isn't -- it only contains scripts and no compilation units, hence no ending up in an (out of tree) build folder. That could change in the future though, so I can see your point.

Reply:
The functional test scripts itself live in the *source* directory. The build directory has a symlink to it. As long as the script in contrib is also symlinked to the build dir, this should not affect out-of-tree builds at all. And all ci scripts use out-of-tree builds, so if this lead to issues, it would be noticed immediately.

Reply:
I've elaborated on why I did this [below](https://github.com/bitcoin/bitcoin/pull/19509#issuecomment-669945057).  If anyone is still against this choice, please let me know!